### PR TITLE
feat(:source): source current ft=lua buffer as Lua code

### LINKED ISSUE #23800: inconsistency with :source-ing lua code from a buffer without file
### Problem

Neovim can source both vimscript and lua code with :source, but only vimscript can be used if the buffer is not backed by a file (i.e. created with :enew and set a proper filetype).

### Steps to reproduce

Consider the following scenario:

```vim
:enew
" write some vim script code
" do not save the buffer
:source " <-- the code gets executed here
```

So for vimscript there is no need to create a file in order to run the code: just make a buffer, write the code, :source, done. Easy.

Let's try the same with lua:

```vim
:enew
:set ft=lua
" write some lua code
" do not save the buffer
:source " <-- prints a bunch of syntax errors for the lua code (as it is not vimscript, obviously)
```

Now do this with that lua buffer:

```vim
:w /tmp/a.lua
:source " <-- lua code gets executed
```

This looks like inconsistent behaviour to me. Although it can be worked around by using "lua << EOF ... EOF", I believe the :source command should work the same way for both languages in all the scenarios

### Expected behavior

The expected behaviour is that :source command runs both vimscript and lua files from buffers without files with filetype set properly.

### Neovim version (nvim -v)

0.9.0

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

macOS 13.3.1

### Terminal name/version

alacritty 0.12.0

### $TERM environment variable

screen-256color

### Installation

homebrew

### PR DESCRIPTION
**Pull request by zeertzjq**
Close #23800

### CODE REVIEW

File: runtime/doc/repeat.txt [hunk 1]
```diff
@@ -183,7 +183,10 @@ For writing a Vim script, see chapter 41 of the user manual |usr_41.txt|.
 
 					*:so* *:source* *load-vim-script*
 :[range]so[urce] [file]	Runs |Ex| commands or Lua code (".lua" files) from
```
Comment (justinmk) [inline runtime/doc/repeat.txt:185]: Just this change would be sufficient, imo:

```suggestion
:[range]so[urce] [file]	Runs |Ex| commands or Lua code (".lua" files or filetype=lua buffers) from
```

File: src/nvim/runtime.c [hunk 2]
```diff
@@ -1926,8 +1926,8 @@ static void cmd_source_buffer(const exarg_T *const eap)
     .buf = ga.ga_data,
     .offset = 0,
   };
-  if (curbuf->b_fname
-      && path_with_extension(curbuf->b_fname, "lua")) {
+  if (strcmp(curbuf->b_p_ft, "lua") == 0
```
Comment (justinmk) [inline src/nvim/runtime.c:1929]: we have `strequal`, fwiw

**Review by justinmk — APPROVED** [review]
Yay!

File: runtime/doc/repeat.txt (hunk 1 above)
Comment (zeertzjq) [reply runtime/doc/repeat.txt:185]: I don't think this is accurate. This PR only changes when behavior when `:source` is called without arguments.

**Pull request merged**